File: drivers/py/ID1000500A.py
import logging, time
from ipdi.ip.pyaip import pyaip, pyaip_init

## IP CONVOLUTIONER driver class
class convolutioner:
    ## Class constructor of IP Convolutioner driver
    #
    # @param self Object pointer
    # @param targetConn Middleware object
    # @param config Dictionary with IP Convolutioner configs
    # @param addrs Network address IP
    def __init__(self, connector, nic_addr, port, csv_file):
        ## Pyaip object
        self.__pyaip = pyaip_init(connector, nic_addr, port, csv_file)

        if self.__pyaip is None:
            logging.debug(error)

        ## Array of strings with information read
        self.dataRX = []

        ## IP Convolutioner IP-ID
        self.IPID = 0

        self.__getID()

        self.__clearStatus()

        logging.debug(f"IP Convolutioner controller created with IP ID {self.IPID:08x}")

    ## Write data in the IP Convolutioner input memory X
    #
    # @param self Object pointer
    # @param data String array with the data to write
    def __writeDataX__(self, data):
        self.sizeDataX = len(data)
        self.__pyaip.writeMem('MMemX', data, len(data), 0)

        logging.debug("Data captured in MemX In")

    ## Write data in the IP Convolutioner input memory Y
    #
    # @param self Object pointer
    # @param data String array with the data to write
    def __writeDataY__(self, data):
        self.sizeDataY = len(data)
        self.__pyaip.writeMem('MMemY', data, len(data), 0)

        logging.debug("Data captured in MemY In")

    ## Read data from the IP Convolutioner output memory
    #
    # @param self Object pointer
    # @param size Amount of data to read
    def __readData__(self, size):
        data = self.__pyaip.readMem('MMemOut', size, 0)
        logging.debug("Data obtained from Mem Out")
        return data

    ## Start processing in IP Convolutioner
    #
    # @param self Object pointer
    def __startIP__(self):
        self.__pyaip.start()

        logging.debug("Start sent")

    ## Set and enable delay in IP Convolutioner processing
    #
    # @param self Object pointer
    # @param msec Delay in millisecond
    def __configurationRegister__(self, dataSizeX, dataSizeY):
        confRegSize = dataSizeX * dataSizeY

        self.__pyaip.writeConfReg('CConfReg', confRegSize, 1, 0)

        logging.debug(f"Configuration Register size is {confRegSize} ms")

    ## Enable IP Convolutioner interruptions
    #
    # @param self Object pointer
    def __enableINT__(self):
        self.__pyaip.enableINT(0, None)

        logging.debug("Int enabled")

    ## Disable IP Convolutioner interruptions
    #
    # @param self Object pointer
    def __disableINT__(self):
        self.__pyaip.disableINT(0)

        logging.debug("Int disabled")

    ## Show IP Convolutioner status
    #
    # @param self Object pointer
    def __status__(self):
        status = self.__pyaip.getStatus()
        logging.info(f"{status:08x}")

        ## Show IP Convolutioner status
        #
        # @param self Object pointer
        def status(self):
            status = self.__pyaip.getStatus()
            logging.info(f"{status:08x}")

    ## Finish connection
    #
    # @param self Object pointer
    def __finish__(self):
        self.__pyaip.finish()

    ## Wait for the completion of the process
    #
    # @param self Object pointer
    def __waitInt__(self):
        waiting = True

        while waiting:

            status = self.__pyaip.getStatus()

            logging.debug(f"status {status:08x}")

            if status & 0x1:
                waiting = False

            time.sleep(0.1)

    ## Get IP ID
    #
    # @param self Object pointer
    def __getID__(self):
        self.IPID = self.__pyaip.getID()

    ## Clear status register of IP Convolutioner
    #
    # @param self Object pointer
    def __clearStatus__(self):
        for i in range(8):
            self.__pyaip.clearINT(i)

    def conv(self, X, Y):
        if len(X) == 0 or len(Y) == 0:
            logging.info("Data X and Data Y can't be empty list")
        if len(X) > 32 or len(Y) > 32:
            logging.info("Data X and Data Y can't be higher than 32")

        self.__writeDataX__(X)
        self.__writeDataY__(Y)

        self.__configurationRegister__(len(X),len(Y))
        self.__startIP__()

        self.__waitInt__()

        result_size = len(X) + len (Y) - 1
        conv_res = self.__readData__(result_size)

        return conv_res


if __name__ == "__main__":
    import sys, random, time, os

    logging.basicConfig(level=logging.DEBUG)
    connector = '/dev/ttyACM0'
    csv_file = '/home/lagisaurio/Documents/Quartus/Convo_New/AIP_Generated/ID1000500A_config.csv'
    addr = 1
    port = 0
    aip_mem_size = 32 #Originalmente 8

    try:
        conv = convolutioner(connector, addr, port, csv_file)
        logging.info("Test Convolutioner: Driver created")
    except:
        logging.error("Test Convolutioner: Driver not created")
        sys.exit()

    random.seed(1)

    X = [random.randrange(2**32) for i in range(0, aip_mem_size)]
    Y = [random.randrange(2 ** 32) for i in range(0, aip_mem_size)]

    try:
        result = conv.conv(X,Y)
        logging.info(f"Convolution result: {[f'{x:08x}' for x in result]}")
    except Exception as e:
        logging.error(f"Error in convlution {str(e)}")
    conv.__finish__()
    logging.info("The End")

File: drivers/c/ID1000500A.c
#include "ID1000500A.h"
#include "caip.h"
#include <stdio.h>
#include <stdbool.h>
#include <string.h> // For memset
#include <stdlib.h> // For malloc

#ifdef _WIN32
#else
#include <unistd.h>
#endif // _WIN32

//Defines
#define INT_DONE    0
#define ONE_FLIT    1
#define ZERO_OFFSET 0
#define STATUS_BITS 8
#define INT_DONE_BIT    0x00000001


/** Global variables declaration (private) */
caip_t      *ID1000500A_aip;
uint32_t    ID1000500A_id = 0;
/*********************************************************************/

/** Private functions declaration */
static uint32_t ID1000500A_getID(uint32_t* id);
static uint32_t ID1000500A_clearStatus(void);
static int32_t ID1000500A_writeData(uint32_t *dataX, uint8_t sizeX, uint32_t *dataY, uint8_t sizeY);
static int32_t ID1000500A_readData(uint32_t *data, uint8_t data_size);
static int32_t ID1000500A_startIP(void);
static int32_t ID1000500A_configurationRegister(uint8_t sizeX, uint8_t sizeY);
static int32_t ID1000500A_waitINT(void);

/* Convolution process*/

//PRIVATE FUNCTIONS

/* Driver initialization*/
int32_t ID1000500A_init(const char *connector, uint8_t nic_addr, uint8_t port, const char *csv_file)
{
    ID1000500A_aip = caip_init(connector, nic_addr, port, csv_file);

    if(ID1000500A_aip == NULL){
        printf("CAIP Object not created");
        return -1;
    }
    ID1000500A_aip->reset();

    ID1000500A_getID(&ID1000500A_id);
    ID1000500A_clearStatus();

    printf("\nIP Convolutioner controller created with IP ID: %08X\n\n", ID1000500A_id);
    return 0;
}

/* Write data MemX*/
int32_t ID1000500A_writeData(uint32_t* dataX, uint8_t sizeX, uint32_t* dataY, uint8_t sizeY)
{
    ID1000500A_aip->writeMem("MMemX", dataX, sizeX, ZERO_OFFSET);

    ID1000500A_aip->writeMem("MMemY", dataY, sizeY, ZERO_OFFSET);
    return 0;
}

/* Read data*/
int32_t ID1000500A_readData(uint32_t *data, uint8_t data_size)
{
    ID1000500A_aip->readMem("MMemOUT", data, data_size, ZERO_OFFSET);
    return 0;
}

/* Start processing*/
int32_t ID1000500A_startIP(void)
{
    ID1000500A_aip->start();
    return 0;
}

/* Configuration Register*/
int32_t ID1000500A_configurationRegister(uint8_t sizeX, uint8_t sizeY)
{
    const uint8_t ADDR_WIDTH = 5;

    unsigned int confRegSize = 0;
    confRegSize = sizeY;
    confRegSize = confRegSize << ADDR_WIDTH;
    confRegSize |= sizeX;
    ID1000500A_aip->writeConfReg("CConfReg", &confRegSize, ONE_FLIT, ZERO_OFFSET);

    return 0;
}

/* Enable interruption notification "Done"*/
int32_t ID1000500A_enableINT(void) {
    ID1000500A_aip->enableINT(INT_DONE, NULL);
    printf("\nINT Done enabled");
    return 0;
}

/* Disable interruption notification "Done"*/
int32_t ID1000500A_disableINT(void)
{
    ID1000500A_aip->disableINT(INT_DONE);
    printf("\nINT Done disabled");
    return 0;
}

/* Show status*/
int32_t ID1000500A_status(void)
{
    uint32_t status;
    ID1000500A_aip->getStatus(&status);
    printf("\nStatus: %08X",status);
    return 0;
}

/* Wait interruption*/
int32_t ID1000500A_waitINT(void)
{
    bool waiting = true;
    uint32_t status;

    while(waiting)
    {
        ID1000500A_aip->getStatus(&status);

        if((status & INT_DONE_BIT)>0)
            waiting = false;

        #ifdef _WIN32
        Sleep(500); // ms
        #else
        sleep(0.1); // segs
        #endif
    }

    ID1000500A_aip->clearINT(INT_DONE);

    return 0;
}

/* Finish*/
int32_t ID1000500A_finish(void)
{
    ID1000500A_aip->finish();
    return 0;
}

uint32_t ID1000500A_getID(uint32_t* id)
{
    ID1000500A_aip->getID(id);

    return 0;
}

uint32_t ID1000500A_clearStatus(void)
{
    for(uint8_t i = 0; i < STATUS_BITS; i++)
        ID1000500A_aip->clearINT(i);

    return 0;
}


void ID1000500A_conv(uint8_t* X, uint8_t sizeX, uint8_t* Y, uint8_t sizeY, uint16_t* result){
    const uint8_t sizeZ = sizeX + sizeY - 1;

    if (sizeX == 0 || sizeY == 0)
        printf("SizeX and SizeY can't be 0");
    if (sizeX > 31 || sizeY > 31)
        printf("SizeX and SizeY can't be higher than 31");


    uint32_t* dataX32 = (uint32_t*)malloc(sizeX * sizeof (uint32_t));
    uint32_t* dataY32 = (uint32_t*)malloc(sizeY * sizeof (uint32_t));
    uint32_t* dataZ32 = (uint32_t*)malloc(sizeZ * sizeof (uint32_t));

    // Inicializar en cero
    memset(dataX32, 0, sizeX);
    memset(dataY32, 0, sizeY);
    memset(dataZ32, 0, sizeZ);

    // Copiar los datos de 8 bits a 32 bits
    for (int i=0; i<sizeX; i++)
        dataX32[i] = X[i];

    for (int i=0; i<sizeY; i++)
        dataY32[i] = Y[i];

    ID1000500A_writeData(dataX32, sizeX, dataY32, sizeY);
    ID1000500A_configurationRegister(sizeX, sizeY);
    ID1000500A_startIP();
    ID1000500A_waitINT();
    ID1000500A_readData(dataZ32, sizeZ);

    for (int i=0; i<sizeZ; i++)
        result[i] = dataZ32[i];
}
